FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1494 @@
+
+static AckInfo_t prvGetAck( uint16_t usPacketId,
+                            bool xRemove )
+{
+    int32_t i = 0;
+    AckInfo_t xFoundAck = { 0 };
+
+    for( i = 0; i < PENDING_ACKS_MAX_SIZE; i++ )
+    {
+        if( pxPendingAcks[ i ].packetId == usPacketId )
+        {
+            xFoundAck = pxPendingAcks[ i ];
+
+            if( xRemove )
+            {
+                pxPendingAcks[ i ].packetId = MQTT_PACKET_ID_INVALID;
+                memset( ( void * ) &( pxPendingAcks[ i ].command ), 0x00, sizeof( Command_t ) );
+            }
+
+            break;
+        }
+    }
+
+    if( xFoundAck.packetId == MQTT_PACKET_ID_INVALID )
+    {
+        LogError( ( "No ack found for packet id %u.", usPacketId ) );
+    }
+
+    return xFoundAck;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvAddSubscription( const char * pTopicFilter,
+                                uint16_t topicFilterLength,
+                                QueueHandle_t pxQueue )
+{
+    int32_t i = 0, availableIndex = SUBSCRIPTIONS_MAX_COUNT;
+
+    /* Start at end of array, so that we will insert at the first available index. */
+    for( i = SUBSCRIPTIONS_MAX_COUNT - 1; i >= 0; i-- )
+    {
+        if( pxSubscriptions[ i ].topicFilterLength == 0 )
+        {
+            availableIndex = i;
+        }
+        else if( ( pxSubscriptions[ i ].topicFilterLength == topicFilterLength ) &&
+                 ( strncmp( pTopicFilter, pxSubscriptions[ i ].pTopicFilter, topicFilterLength ) == 0 ) )
+        {
+            /* If a subscription already exists, don't do anything. */
+            if( pxSubscriptions[ i ].pResponseQueue == pxQueue )
+            {
+                LogWarn( ( "Subscription already exists. " ) );
+                availableIndex = SUBSCRIPTIONS_MAX_COUNT;
+                break;
+            }
+        }
+    }
+
+    if( availableIndex < SUBSCRIPTIONS_MAX_COUNT )
+    {
+        pxSubscriptions[ availableIndex ].topicFilterLength = topicFilterLength;
+        pxSubscriptions[ availableIndex ].pResponseQueue = pxQueue;
+        memcpy( pxSubscriptions[ availableIndex ].pTopicFilter, pTopicFilter, topicFilterLength );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvRemoveSubscription( const char * pTopicFilter,
+                                   uint16_t topicFilterLength )
+{
+    int32_t i = 0;
+
+    for( i = 0; i < SUBSCRIPTIONS_MAX_COUNT; i++ )
+    {
+        if( pxSubscriptions[ i ].topicFilterLength == topicFilterLength )
+        {
+            if( strncmp( pxSubscriptions[ i ].pTopicFilter, pTopicFilter, topicFilterLength ) == 0 )
+            {
+                pxSubscriptions[ i ].topicFilterLength = 0;
+                pxSubscriptions[ i ].pResponseQueue = NULL;
+                memset( ( void * ) pxSubscriptions[ i ].pTopicFilter, 0x00, sizeof( pxSubscriptions[ i ].pTopicFilter ) );
+            }
```

Comment:
Repeating the same question about multiple tasks subscribing to one topic filter, wouldn't it be better to check for the Queue handle as well and remove only that?

Reply:
That's how this function originally was, but I removed that parameter when I made it so that multiple unsubscribes could be sent in a single packet. If we're unsubscribing from three topics, but one is being used by another task, we'd need to do some editing in the buffer

Reply:
Ok Sounds good. Thanks